### PR TITLE
[rocjitsu] Native DRM ioctls; harden daemon-mode KFD lifecycle

### PR DESCRIPTION
**Pull request by atgutier**
## Motivation

The rocjitsu interposer layer intercepts at the system call level (to be precise, the C lib's syscall wrappers) and it doesn't intercept any user-mode driver API explicitly. As a temporary workaround we intercepted some libdrm amdgpu calls for expediency. This PR removes those cases in favor of native support for amdgpu DRM ioctls, etc.

## Technical Details

Replace the libdrm library-symbol shims with native handling of the amdgpu/DRM ioctl ABI so real libdrm_amdgpu runs unmodified over the simulator, and harden the existing daemon mode for multi-process clients with cross-process memory access, client-PID tracking, and process reuse.

Serve the amdgpu/DRM driver natively (interposer.cpp):
  - Remove all libdrm symbol interceptions (amdgpu_device_initialize/2, amdgpu_device_deinitialize/get_fd, amdgpu_query_gpu_info, amdgpu_get_marketing_name, amdgpu_bo_import/export/free/cpu_map/ cpu_unmap/va_op/query_info/set_metadata, drmCommandWriteRead, drmGetDevice(s)/drmFreeDevice(s)) and the dlsym override; real libdrm_amdgpu now runs and its calls are caught at the syscall layer
  - Handle DRM_IOCTL_AMDGPU_INFO for the queries libdrm issues during amdgpu_device_initialize: ACCEL_WORKING, DEV_INFO (with VA aperture), VRAM_GTT, MEMORY, and READ_MMR_REG; VERSION and PRIME_FD_TO_HANDLE remain handled
  - Interpose fcntl64 (libdrm imports fcntl64@GLIBC_2.28, not fcntl) so F_DUPFD_CLOEXEC on the render fd propagates DRM tracking; ioctls on the duped fd are otherwise unrecognized
  - Redirect individual /dev/dri/renderD*,card* node paths (not just the /dev/dri directory) into the synthetic dev_dri tree so libdrm's drmGetMinorType access() probe classifies every GPU's render node on multi-GPU configs
  - Drop the RjFakeBo/dmabuf bridge, Rj* libdrm mirror types, and the dlsym passthrough machinery left dead by the removals

Vendor kernel DRM UAPI headers (MIT):
  - Add external_headers/drm_headers/linux/uapi/drm/{drm.h,amdgpu_drm.h} (libdrm 2.4.125) and a DRM_INCLUDE_DIR include path, so the ioctl handlers use the real kernel structs (drm_amdgpu_info, drm_amdgpu_info_device, drm_amdgpu_info_vram_gtt, drm_amdgpu_memory_info, drm_version) rather than hand-mirrored copies

Harden daemon-mode KFD lifecycle, memory bridging, and process reuse (daemon mode and its RemoteDriver/RPC transport already exist; this adds the multi-process client support on top):
  - Interposer: remote_mutex_ guards RemoteDriver create/destroy, remote_kfd_fd_ made atomic, close_remote()/close_remote_fd_only() for clean shutdown, AMDKFD_IOCTL_BASE fallback for late ioctls, reset_after_fork() clears all containers and reinits remote_mutex_
  - SimulatedDriver: client_pid parameter on open_process() with process reuse, open_process() TOCTOU fix (pid search + create under one lock), unmap_memory_ioctl unmaps GPU pages without erasing the allocation record, set_process_client_pid() for post-hoc pid attachment
  - KfdProcess: client_pid_ plus retain_open()/release_open() refcount
  - gpu_memory.h: process_vm_readv/writev fallback for cross-process access, per-VMID client_pid tracking, warnings on access failure
  - remote_driver.cpp: extract dmabuf fd from EXPORT_DMABUF RPC response
  - rj_vm.h/cpp: rj_vm_device_open_for_client_pid() and rj_vm_device_set_client_pid() APIs
  - main.cpp: extract client PID via SO_PEERCRED, pass to rj_vm_device_open_for_client_pid()

## JIRA ID

N/A

## Test Plan

Run all CI tests, run OPT125, run hrx-systems tests.

## Test Result

CI tests: PASSED
OPT125: PASSED with 100% accurate tokens/logits compared to CPU reference run.
hrs-systems tests: 61/107 PASSED (known failures, no regressions)

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by atgutier** [comment]
@copilot do another review

**Comment by atgutier** [comment]
@copilot review again

**Review by kuhar — COMMENTED** [review]
It would be nice to add a small amount of targeted coverage for the new daemon/libdrm surface. The important, practical tests would be:

1. A libdrm smoke test that runs `amdgpu_device_initialize` / `amdgpu_query_gpu_info` through the interposer. This is the headline behavior and the init cascade is field/order-sensitive. The existing `daemon_test.cpp` helper pattern looks suitable; if linking a libdrm helper is too heavy, factor the AMDGPU_INFO DEV_INFO filling into a helper and unit-test that path.
2. Retarget `drm_info_layout_test.cpp` to the vendored `drm_amdgpu_info_device` struct and assert the fields populated by DEV_INFO, including the virtual-address and high-VA fields. The current `LocalDevInfo` mirror no longer corresponds to production code.
3. Add one focused daemon/process lifecycle test for same-client `open_process(client_pid)` reuse and last-close teardown. This is easy in `KfdIoctlTest` and also guards the open/close refcount behavior added here.

Cross-process `process_vm_readv/writev` coverage would be useful, but it is a second-tier/Linux-environment-sensitive test. I would not block this PR on that alone if the smoke and lifecycle coverage above are added.

Agent comments below.

The new /dev/dri redirect should also be applied in the 64-bit and versioned stat wrappers. stat(), lstat(), access(), and opendir() now call redirect_dev_dri(), but stat64(), lstat64(), __xstat*(), and __lxstat*() still only check the sysfs/dev-char redirects. A libdrm probe using those entry points can still hit the host /dev/dri/renderD* and miss the synthetic render node.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp [hunk 1]
```diff
@@ -811,8 +709,8 @@ int openat64(int dirfd, const char *path, int flags, ...) {
 
 int close(int fd) {
   assert(InterposerContext::real.ready());
-  if (auto *remote = InterposerContext::ctx.remote_lookup(fd))
-    return remote->close();
+  if (InterposerContext::ctx.remote_lookup(fd))
+    return InterposerContext::ctx.close_remote_fd_only();
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp]: We need the last daemon KFD close to tear down the remote driver, not just the synthetic fd. The close() path for remote_lookup(fd) calls close_remote_fd_only(), while close_remote()/try_close_remote() have no call sites and the destructor is empty. That means RemoteDriver::close()/RPC_CLOSE are not sent on /dev/kfd close, leaving daemon-side process state alive until process exit. Please ref-count the primary and dup-tracked KFD fds and call RemoteDriver::close() on the last close.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/simulated_driver.cpp [hunk 2]
```diff
@@ -364,12 +376,25 @@ uint32_t SimulatedDriver::open_process() {
   uint32_t pid;
   {
     std::lock_guard<std::mutex> lk(process_mutex_);
+    if (client_pid > 0) {
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/simulated_driver.cpp]: We should keep the client-pid reuse and close-refcount conditions symmetric. open_process() reuses and calls retain_open() for any client_pid > 0, but close() only calls release_open() when daemon_mode_ is true. If the public API is called with a nonzero client PID outside daemon mode, the first close can tear down a still-retained process. Gate reuse on daemon_mode_ too, or remove the daemon-mode guard from close; a focused same-client open/open/close/close test would lock this down.

File: emulation/rocjitsu/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,23 @@ All notable changes to rocjitsu are documented here.
 Format follows [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 Versioning follows [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.2.0] - 2026-06-23
```
Comment (kuhar) [inline emulation/rocjitsu/CHANGELOG.md:7]: We have a bunch of recent changes that should probably also be added here... Perhaps worth sending a reminder on Teams before we bump to 0.2.0?

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h [hunk 4]
```diff
@@ -126,6 +134,9 @@ class GpuMemory : public simdojo::SparseMemory {
   uint8_t read8(uint64_t addr, uint32_t vmid = 0) const {
     if (auto *p = translate(addr, vmid))
       return p[addr & PAGE_MASK];
+    uint8_t val = 0;
+    if (vmid > 0 && read_client_memory(addr, &val, 1, vmid))
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h:138]: nit / optional: in c++ >=17 you could define `uint8_t val = 0;` inside the if condition itself. Here and elsewhere below.

File: emulation/rocjitsu/CHANGELOG.md (hunk 3 above)
Comment (atgutier) [reply emulation/rocjitsu/CHANGELOG.md:7]: I have only been adding changes to this log that directly affect the versioning. Are you suggesting to log all changes here?

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp (hunk 1 above)
Comment (atgutier) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp]: I had another commit that did this, so I just added it to this PR.

LMK if you're comments have been addressed.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h (hunk 4 above)
Comment (atgutier) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h:138]: Will cleanup stuff like this in a targeted refactor/cleanup. I have a list going...

File: emulation/rocjitsu/CHANGELOG.md (hunk 3 above)
Comment (kuhar) [reply emulation/rocjitsu/CHANGELOG.md:7]: I see, I thought this was for major user-facing changes and, for example, supported targets would go in here too

Comment (atgutier) [reply emulation/rocjitsu/CHANGELOG.md:7]: Yes, that is the right idea. We can start that from now on. Technically, those should also bump the minor version. I think once we get to 1.0.0 we will be far more strict on this.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp [hunk 5]
```diff
@@ -913,6 +927,15 @@ int ioctl(int fd, unsigned long request, ...) {
     if (auto *remote = InterposerContext::ctx.remote_lookup(InterposerContext::ctx.remote_kfd_fd()))
       return remote->ioctl(request, arg);
   }
+  // Late-ioctl safety net: an AMDKFD ('K') ioctl may arrive on a tracked KFD fd
+  // whose primary remote handle changed underneath it (e.g. a close/dup race in
+  // daemon mode). Forward only AMDKFD-typed ioctls, and only on fds we already
+  // track as KFD (primary or dup), so an arbitrary unrelated fd carrying a
+  // type-'K' ioctl is never misrouted to the remote KFD driver.
+  if (_IOC_TYPE(request) == AMDKFD_IOCTL_BASE && InterposerContext::ctx.is_kfd_tracked(fd)) {
+    if (auto *remote = InterposerContext::ctx.remote())
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp:962]: We should make `remote_` access consistently synchronized. `teardown_remote()` deletes and clears `remote_` under `remote_mutex_`, but `remote()`, `remote_lookup()`, the AMDKFD fallback, and the mmap path read the raw pointer without that mutex. The open refcount lowers the practical use-after-free risk, but it does not make the pointer read/write data-race-free. Either load/store `remote_` through an atomic pointer or take `remote_mutex_` in the read helpers.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h [hunk 6]
```diff
@@ -214,6 +243,42 @@ class GpuMemory : public simdojo::SparseMemory {
...
+    return (it != vmid_table_.end()) ? it->second.client_pid : 0;
+  }
+
+  bool read_client_memory(uint64_t addr, void *dst, size_t len, uint32_t vmid) const {
+    pid_t pid = client_pid_for_vmid(vmid);
+    if (pid <= 0)
+      return false;
+    iovec local{dst, len};
+    iovec remote{reinterpret_cast<void *>(addr), len};
+    ssize_t rc = process_vm_readv(pid, &local, 1, &remote, 1, 0);
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/vm/amdgpu/gpu_memory.h:258]: We should make the daemon cross-process memory contract explicit. `GpuMemory` now uses `process_vm_readv/writev` for `vmid > 0` misses, but on systems with Yama `ptrace_scope=1` a daemon child or unrelated attach process cannot read the client without ptrace permission; this runner has that setting. Today failure only logs and falls back to `SparseMemory`, so a real client-memory read can become zero or stale data. Please arrange the ptrace permission relationship, or use a shared-memory path, and treat `EPERM`/short transfer as a daemon-mode error instead of a cache miss.

File: emulation/rocjitsu/lib/rocjitsu/include/rocjitsu/vm/rj_vm.h [hunk 7]
```diff
@@ -223,10 +224,14 @@ RJ_API_EXPORT rj_status_t rj_vm_execute(rj_vm_t *vm, rj_vm_cmd_t *cmd);
...
+/// @brief Open the VM's simulated device and create (or reuse) a KFD process.
 /// @param[in] vm VM handle.
-/// @param[out] process_id The new process ID (may be NULL for legacy callers).
-RJ_API_EXPORT rj_status_t rj_vm_device_open(rj_vm_t *vm, uint32_t *process_id);
+/// @param[in] client_pid Connecting client's OS PID for daemon mode (enables
+///   cross-process memory access and process reuse); pass 0 in local mode where
+///   there is no separate client process. If a process already exists for a
+///   nonzero client_pid, it is reused.
+/// @param[out] process_id The simulator's process handle (may be NULL).
+RJ_API_EXPORT rj_status_t rj_vm_device_open(rj_vm_t *vm, pid_t client_pid, uint32_t *process_id);
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/include/rocjitsu/vm/rj_vm.h]: We should keep the public C API type-stable here. `rj_vm.h` otherwise exposes fixed-width types or project handles, but this change adds `<sys/types.h>` and `pid_t` to `rj_vm_device_open`, making the common VM header POSIX-specific and ABI-width-dependent. Please use a project typedef or fixed-width client PID type in the exported API, and translate to `pid_t` inside the Linux daemon path.

**Review by kuhar — COMMENTED** [review]
One focused daemon lifecycle regression test would be worthwhile here, but I would not ask for a full libdrm/ioctl test stack in this PR.

The cheap path is in `KfdIoctlTest`: it already constructs a daemon-mode `SimulatedDriver` directly, and the file already has local KFD open-refcount coverage. A targeted test could call `open_process(client_pid)` twice with the same nonzero PID, assert the same process ID is reused, close it once, verify a process-scoped ioctl still succeeds, then close it again and verify the process is gone. That covers the client-PID reuse/refcount behavior added here without launching the daemon, linking libdrm, or refactoring `interposer.cpp`.

A real libdrm smoke test for `amdgpu_device_initialize` / `AMDGPU_INFO_DEV_INFO` would be useful follow-up coverage if we add optional libdrm test plumbing, but I would not make that a blocker for this PR.

Original anchor: `emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp:965`

`dup2`/`dup3` should not mutate KFD tracking before the syscall succeeds. The current wrappers call `untrack_dup(newfd)` first, so `dup2(fd, fd)` drops tracking before the POSIX no-op and `dup3(fd, fd, ...)` drops it before the expected `EINVAL`. That can decrement the open ref while the descriptor remains live. Capture the old tracking state, skip the `dup2(fd, fd)` no-op case, and only replace tracking after a successful duplication/replacement.

Original anchor: `emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/kmd/linux/interposer.cpp:803`

`DRM_IOCTL_VERSION` can copy one byte past a caller buffer. The guard accepts `name_len >= sizeof("amdgpu") - 1`, but the copy uses `sizeof(drv_name)`, including the NUL byte. A caller that allocates the exact length returned by the query path gets a 7-byte write into a 6-byte buffer. Either require space for the terminator or copy only the advertised length and NUL-terminate only when there is room.

**Review by kuhar — APPROVED** [review]
Overall LGTM, none of the agent issues posted above seem blocking

File: emulation/rocjitsu/lib/rocjitsu/include/rocjitsu/vm/rj_vm.h (hunk 7 above)
Comment (atgutier) [reply emulation/rocjitsu/lib/rocjitsu/include/rocjitsu/vm/rj_vm.h]: Yeah, this is actually a big deal. I will fix this.
